src/Npgsql/NpgsqlDataSourceBuilder.cs

```diff
@@ -232,6 +232,30 @@ public NpgsqlDataSourceBuilder UseRootCertificateCallback(Func<X509Certificate2>
         return this;
     }
 
+    /// <summary>
+    /// Configures a password provider, which is called by the data source when opening connections.
+    /// </summary>
+    /// <param name="passwordProvider">
+    /// A callback that may be invoked during <see cref="NpgsqlConnection.Open()" /> which returns the password to be sent to PostgreSQL.
+    /// </param>
+    /// <param name="passwordProviderAsync">
+    /// A callback that may be invoked during <see cref="NpgsqlConnection.OpenAsync(CancellationToken)" /> which returns the password to be sent to PostgreSQL.
+    /// </param>
+    /// <returns>The same builder instance so that multiple calls can be chained.</returns>
+    /// <remarks>
+    /// <para>
+    /// The provided callback is invoked when opening connections. Therefore its important the callback internally depends on cached
+    /// data or returns quickly otherwise. Any unnecessary delay will affect connection opening time.
+    /// </para>
+    /// </remarks>
+    public NpgsqlDataSourceBuilder UsePasswordProvider(
+        Func<NpgsqlConnectionStringBuilder, string>? passwordProvider,
```

Comment:
nit: same naming nit, passwordProviderSync?

Reply:
Done to align with naming used in UseConnectionInitializer